File: ts/src/integrations/_shared/sink.ts
/**
 * TraceSink interface + FileSink implementation (shared across integrations).
 *
 * Originally shipped under `ts/src/integrations/openai/sink.ts` (A2-II-b);
 * lifted here to be consumed by every provider integration.
 *
 * No beforeExit by default; `registerBeforeExit: true` opts in.
 */
import { appendFileSync, fsyncSync, mkdirSync, openSync, closeSync } from "node:fs";
import { dirname } from "node:path";

export interface TraceSink {
  add(trace: Record<string, unknown>): void;
  flush(): void;
  close(): void;
}

export interface FileSinkOptions {
  /** Max traces before auto-flush. Default: 64. */
  batchSize?: number;
  /** Seconds since last flush before auto-flush. Default: 5. */
  flushIntervalSeconds?: number;
  /** How to handle write errors. Default: "raise". */
  onError?: "raise" | "log-and-drop";
  /** Register a process.on("beforeExit") closer. Default: false. */
  registerBeforeExit?: boolean;
}

/** Batched JSONL trace sink.
 *
 * Buffers traces in memory; flushes on batchSize or flushIntervalSeconds elapsed.
 * Writes are append-only with fsync. Mirror of Python FileSink.
 */
export class FileSink implements TraceSink {
  private readonly _path: string;
  private readonly _batchSize: number;
  private readonly _flushIntervalMs: number;
  private readonly _onError: "raise" | "log-and-drop";
  private _buffer: Array<Record<string, unknown>> = [];
  private _lastFlushAt: number = Date.now();
  private _closed = false;

  constructor(path: string, options?: FileSinkOptions) {
    this._path = path;
    this._batchSize = options?.batchSize ?? 64;
    this._flushIntervalMs = (options?.flushIntervalSeconds ?? 5.0) * 1000;
    this._onError = options?.onError ?? "raise";
    if (options?.registerBeforeExit) {
      process.on("beforeExit", () => {
        try { this.close(); } catch { /* best-effort */ }
      });
    }
  }

  add(trace: Record<string, unknown>): void {
    if (this._closed) {
      throw new Error("FileSink is closed");
    }
    this._buffer.push(trace);
    if (this._buffer.length >= this._batchSize) {
      this._flushLocked();
      return;
    }
    if (Date.now() - this._lastFlushAt >= this._flushIntervalMs) {
      this._flushLocked();
    }
  }

  flush(): void {
    if (!this._closed) {
      this._flushLocked();
    }
  }

  close(): void {
    if (this._closed) return;
    this._flushLocked();
    this._closed = true;
  }

  private _flushLocked(): void {
    if (this._buffer.length === 0) {
      this._lastFlushAt = Date.now();
      return;
    }
    try {
      mkdirSync(dirname(this._path), { recursive: true });
      const lines = this._buffer
        .map((t) => JSON.stringify(t, _sortedReplacer))
        .join("\n") + "\n";
      appendFileSync(this._path, lines, "utf-8");
      // fsync the file to ensure durability
      const fd = openSync(this._path, "r");
      try {
        fsyncSync(fd);
      } finally {
        closeSync(fd);
      }
    } catch (err) {
      if (this._onError === "raise") throw err;
      // log-and-drop
      try {
        process.stderr.write(`[FileSink] flush failed: ${String(err)}\n`);
      } catch { /* ignore */ }
    } finally {
      this._buffer = [];
      this._lastFlushAt = Date.now();
    }
  }
}

/** JSON replacer that produces sorted-key output (matches Python sort_keys=True). */
function _sortedReplacer(_key: string, value: unknown): unknown {
  if (value !== null && typeof value === "object" && !Array.isArray(value)) {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).sort(([a], [b]) => a.localeCompare(b)),
    );
  }
  return value;
}

File: autocontext/src/autocontext/integrations/_shared/sink.py
"""TraceSink protocol + FileSink implementation (shared across integrations).

Originally shipped under ``autocontext.integrations.openai._sink`` (A2-II-b);
lifted here to be consumed by every provider integration (openai, anthropic, …).

No atexit by default; ``register_atexit=True`` opts in.
"""
from __future__ import annotations

import atexit
import json
import logging
import os
import time
from pathlib import Path
from threading import Lock
from typing import Any, Literal, Protocol, runtime_checkable

_logger = logging.getLogger("autocontext.integrations._shared.FileSink")


@runtime_checkable
class TraceSink(Protocol):
    def add(self, trace: dict[str, Any]) -> None: ...
    def flush(self) -> None: ...
    def close(self) -> None: ...


class FileSink:
    """Batched JSONL trace sink.

    Buffers traces in memory; flushes on ``batch_size`` or ``flush_interval_seconds``
    elapsed since the last write (whichever comes first). Writes are append-only
    with fsync.
    """

    def __init__(
        self,
        path: str | Path,
        *,
        batch_size: int = 64,
        flush_interval_seconds: float = 5.0,
        on_error: Literal["raise", "log-and-drop"] = "raise",
        register_atexit: bool = False,
    ) -> None:
        self._path = Path(path)
        self._batch_size = batch_size
        self._flush_interval_seconds = flush_interval_seconds
        self._on_error = on_error
        self._buffer: list[dict[str, Any]] = []
        self._lock = Lock()
        self._last_flush_at = time.monotonic()
        self._closed = False
        if register_atexit:
            atexit.register(self._atexit_handler)

    def add(self, trace: dict[str, Any]) -> None:
        with self._lock:
            if self._closed:
                raise RuntimeError("FileSink is closed")
            self._buffer.append(trace)
            if len(self._buffer) >= self._batch_size:
                self._flush_locked()
                return
            if time.monotonic() - self._last_flush_at >= self._flush_interval_seconds:
                self._flush_locked()

    def flush(self) -> None:
        with self._lock:
            self._flush_locked()

    def close(self) -> None:
        with self._lock:
            if self._closed:
                return
            self._flush_locked()
            self._closed = True

    def _flush_locked(self) -> None:
        if not self._buffer:
            self._last_flush_at = time.monotonic()
            return
        try:
            self._path.parent.mkdir(parents=True, exist_ok=True)
            with self._path.open("a", encoding="utf-8") as f:
                for trace in self._buffer:
                    f.write(json.dumps(trace, separators=(",", ":"), sort_keys=True))
                    f.write("\n")
                f.flush()
                os.fsync(f.fileno())
        except OSError as exc:
            if self._on_error == "raise":
                raise
            _logger.warning("FileSink flush failed: %s", exc)
        finally:
            self._buffer.clear()
            self._last_flush_at = time.monotonic()

    def _atexit_handler(self) -> None:
        try:
            self.close()
        except Exception:  # pragma: no cover
            pass
